kettle-plugins/kafka/src/main/java/org/pentaho/big/data/kettle/plugins/kafka/KafkaStreamSource.java

```diff
@@ -0,0 +1,186 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.big.data.kettle.plugins.kafka;
+
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.errors.WakeupException;
+import org.pentaho.di.core.row.RowDataUtil;
+import org.pentaho.di.core.row.ValueMetaInterface;
+import org.pentaho.di.core.variables.VariableSpace;
+import org.pentaho.di.trans.streaming.common.BlockingQueueStreamSource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.IntStream;
+
+public class KafkaStreamSource extends BlockingQueueStreamSource<List<Object>> {
+
+  private static Class<?> PKG = KafkaConsumerInputMeta.class; // for i18n purposes, needed by Translator2!!   $NON-NLS-1$
+  private final Logger logger = LoggerFactory.getLogger( getClass() );
+
+  private final VariableSpace variables;
+  private KafkaConsumerInputMeta kafkaConsumerInputMeta;
+  private KafkaConsumerInputData kafkaConsumerInputData;
+  private Map<KafkaConsumerField.Name, Integer> positions;
+  private final HashMap<TopicPartition, Long> maxOffsets = new HashMap<>();
+
+  private Consumer consumer;
+  private final ExecutorService executorService = Executors.newCachedThreadPool();
+  private KafkaConsumerCallable callable;
+  private Future<Void> future;
+
+  public KafkaStreamSource( Consumer consumer, KafkaConsumerInputMeta inputMeta,
+                            KafkaConsumerInputData kafkaConsumerInputData, VariableSpace variables,
+                            KafkaConsumerInput kafkaStep ) {
+    super( kafkaStep );
+    positions = new HashMap<>();
+    this.consumer = consumer;
+    this.variables = variables;
+    this.kafkaConsumerInputData = kafkaConsumerInputData;
+    this.kafkaConsumerInputMeta = inputMeta;
+  }
+
+  @Override public void close() {
+    super.close();
+    callable.shutdown();
+    future.cancel( true );
+  }
+
+  @Override public void open() {
+    if ( future != null ) {
+      // TODO: create message property
```

Comment:
Wasn't sure if we want to create a specific message property

Reply:
I don't think this one needs to be localized

Reply:
👍